src/gui/sidebar/previews/page/SidebarPreviewPages.cpp

```diff
@@ -11,13 +11,53 @@
 
 SidebarPreviewPages::SidebarPreviewPages(Control* control, GladeGui* gui, SidebarToolbar* toolbar)
  : SidebarPreviewBase(control, gui, toolbar)
+ , contextMenu(gui->get("sidebarPreviewContextMenu"))
 {
 	XOJ_INIT_TYPE(SidebarPreviewPages);
+
+	// Connect the context menu actions
+	const std::map<std::string, SidebarActions> ctxMenuActions = {
+	        {"sidebarPreviewDuplicate", SIDEBAR_ACTION_COPY},
+	        {"sidebarPreviewDelete", SIDEBAR_ACTION_DELETE},
+	        {"sidebarPreviewMoveUp", SIDEBAR_ACTION_MOVE_UP},
+	        {"sidebarPreviewMoveDown", SIDEBAR_ACTION_MOVE_DOWN},
+	        {"sidebarPreviewNewBefore", SIDEBAR_ACTION_NEW_BEFORE},
+	        {"sidebarPreviewNewAfter", SIDEBAR_ACTION_NEW_AFTER},
+	};
+
+	for (const auto& pair: ctxMenuActions)
+	{
+		GtkWidget* const entry = gui->get(pair.first);
+		g_assert(entry != nullptr);
+
+		// Unfortunately, we need a fairly complicated mechanism to keep track
+		// of which action we want to execute.
+		typedef SidebarPreviewPages::ContextMenuData Data;
+		auto userdata = std::unique_ptr<Data>(new Data{this->toolbar, pair.second});
```

Comment:
use `auto userdata = mem::make_unique<Data>(this->toolbar, pair.second);`

Reply:
That doesn't work because `Data` is a struct, and trying to work around it is cumbersome.

Reply:
Just pass it as struct then, it will compile to the same assembler code.
Checkout this: https://www.godbolt.org/z/nweON9
It's a bit shorter and you don't need to use new :D.